### PR TITLE
[RDY] Dispersion round 2

### PR DESCRIPTION
**Pull request by mutability**
@mugling: Here's the PR you wanted. What are the things you want to see before you would accept this?

### CODE REVIEW

File: src/creature.h [hunk 1]
```diff
@@ -185,6 +185,11 @@ class Creature
         dealt_projectile_attack projectile_attack( const projectile &proj, const tripoint &target,
                                                    double total_dispersion );
 
+        /**
+         * Return the chance that a projectile attack will hit with at least the given accuracy
```
Comment (mugling) [inline src/creature.h]: What are the bounds for chance?
Function comments should be doxygen style

File: src/monattack.cpp [hunk 2]
```diff
@@ -514,7 +514,7 @@ bool mattack::acid_accurate(monster *z)
     proj.proj_effects.insert( "NO_DAMAGE_SCALING" );
     proj.impact.add_damage( DT_ACID, rng( 3, 5 ) );
     // Make it arbitrarily less accurate at close ranges
-    z->projectile_attack( proj, target->pos(), rng( 0, 8000 / range ) );
+    z->projectile_attack( proj, target->pos(), 8000 / range );
```
Comment (mugling) [inline src/monattack.cpp]: You might want to review this forum post about this not being updated in the aiming refactor and therefore being to inaccurate. Consider replacement with a range independent constant.

File: src/ranged.cpp [hunk 3]
```diff
@@ -71,15 +71,98 @@ size_t blood_trail_len( int damage )
     return 0;
 }
 
+// projectile_attack handles a single shot aimed at
```
Comment (mugling) [inline src/ranged.cpp]: What is this wordwrapped to? col 100 would be reasonable. Less vertical paging is a good thing for code clarity.

File: src/ranged.cpp [hunk 4]
```diff
@@ -71,15 +71,98 @@ size_t blood_trail_len( int damage )
...
+// dispersion_sigmas and occupied_tile_fraction can be
+// tuned together to affect the spread of misses without
+// affecting hit or hit quality chances, e.g. if you
+// increase dispersion_sigmas by 10% and also decrease
+// occupied_tile_fraction by 10%, then the cone of misses
+// will get 10% tighter but the overall hit and hit
+// quality chances are unchanged.
+
+// these parameters balance the hit/miss chance:
+const double dispersion_sigmas = 2.4;      // how many standard deviations does "dispersion" represent?
```
Comment (mugling) [inline src/ranged.cpp]: Magic numbers need either justification or alternatively a declaration that 'I found it like this and nobody knows why'. You should move your rationale from the above preamble to the variable documentation.

File: src/ranged.cpp [hunk 5]
```diff
@@ -71,15 +71,98 @@ size_t blood_trail_len( int damage )
...
+const double occupied_tile_fraction = 0.5; // how much of the target tile is occupied by the target?
```
Comment (mugling) [inline src/ranged.cpp]: Prefer doxygen comments `/** foo */` for global constants.

Also consider putting them in `game_constants.h`, at the top of the relevant header or within the consuming function. Buried in the middle of an implementation file doesn't make their existence obvious.

I'd prefer all constants in one file (so it's easy to find what the adjustments to certain algorithms can be) but I know @Coolthulhu has a different view so you want to wait for his opinion on that latter part.

File: src/ranged.cpp [hunk 6]
```diff
@@ -71,15 +71,98 @@ size_t blood_trail_len( int damage )
...
 
-    // an isosceles triangle is formed by the intended and actual targets
-    double missed_by = iso_tangent( range, dispersion );
+    // dispersion is a measure of the dispersion of shots due to the gun + shooter characteristics
+    // i.e. it is independent of any particular shot
+
+    // shot_dispersion is the actual dispersion for this particular shot, i.e.
+    // the error angle between where the shot was aimed and where this one actually went
+    // NB: some cases pass dispersion == 0 for a "never misses" shot e.g. bio_magnet,
+    double shot_dispersion = ( dispersion > 0 ? normal_roll( 0.0, dispersion / dispersion_sigmas ) : 0 );
```
Comment (mugling) [inline src/ranged.cpp]: No spurious parenthesis for ternary conditionals. It's not consistent with other code an barring rare exceptions is generally bad for clarity.

File: src/ranged.cpp [hunk 7]
```diff
@@ -71,15 +71,98 @@ size_t blood_trail_len( int damage )
...
+
+    // number of tiles we missed by
```
Comment (mugling) [inline src/ranged.cpp]: Spurious comment considering the variable is named below

**Comment by mutability** [comment]
This also fixes #19013, #19014, #19015, #19016, #19017, #19018

File: src/ranged.cpp [hunk 8]
```diff
@@ -118,17 +201,17 @@ dealt_projectile_attack Creature::projectile_attack( const projectile &proj_arg,
...
         // We missed enough to target a different tile
         double dx = target_arg.x - source.x;
         double dy = target_arg.y - source.y;
         double rad = atan2( dy, dx );
-        // Cap spread at 30 degrees or it gets wild quickly
-        double spread = std::min( ARCMIN( dispersion ), DEGREES( 30 ) );
-        rad += rng_float( -spread, spread );
+
+        // cap wild misses at +/- 30 degrees
+        rad += std::max( std::min( ARCMIN( shot_dispersion ), DEGREES( 30 ) ), DEGREES( -30 ) );
```
Comment (mugling) [inline src/ranged.cpp]: Tangent: We really ought to implement a `clamp` function/macro

File: src/ranged.cpp [hunk 9]
```diff
@@ -341,29 +424,52 @@ dealt_projectile_attack Creature::projectile_attack( const projectile &proj_arg,
     return attack;
 }
 
+static double erfinv( double x )
```
Comment (mugling) [inline src/ranged.cpp]: You need to unit test this somehow. Checking the output is commutative with inverse function from the C++ standard library might be one option?

File: src/ranged.cpp [hunk 10]
```diff
@@ -341,29 +424,52 @@ dealt_projectile_attack Creature::projectile_attack( const projectile &proj_arg,
...
+{
+    const double epsilon = 1e-6;
+    double z = 0.0;
+
+    // find erf(z) - x = 0
+    // d/dz ( erf(z) - x ) = 2/sqrt(pi) . e^(-z^2)
+
+    for( int n = 0; n < 10; ++n ) {
```
Comment (mugling) [inline src/ranged.cpp]: I've not reviewed the maths yet but this still appears to be iterative?

File: src/ranged.cpp [hunk 11]
```diff
@@ -341,29 +424,52 @@ dealt_projectile_attack Creature::projectile_attack( const projectile &proj_arg,
...
 {
     if( !gun.is_gun() || !gun.ammo_sufficient() ) {
         return 0;
     }
 
+    double max_range = gun.gun_range( this );
+
     if( chance == 0 ) {
-        return gun.gun_range( this );
+        return max_range;
```
Comment (mugling) [inline src/ranged.cpp]: No need to change this line

File: src/ranged.cpp [hunk 12]
```diff
@@ -341,29 +424,52 @@ dealt_projectile_attack Creature::projectile_attack( const projectile &proj_arg,
...
+    return z;
+}
+
 double player::gun_current_range( const item& gun, double penalty, unsigned chance, double accuracy ) const
 {
     if( !gun.is_gun() || !gun.ammo_sufficient() ) {
         return 0;
     }
 
+    double max_range = gun.gun_range( this );
```
Comment (mugling) [inline src/ranged.cpp]: Useless constants only increase the line count of your PR and therefore the effort involved in review

**Comment by mutability** [comment]
nb:
```
$ ./tests/cata_test gun_aiming
Finalizing and verifying
===============================================================================
All tests passed (2704 assertions in 1 test case)
```

File: src/ranged.cpp [hunk 13]
```diff
@@ -929,33 +1035,50 @@ static int do_aim( player &p, const std::vector<Creature *> &t, int cur_target,
...
-    // Confidence is chance of the actual shot being under the target threshold,
-    // This simplifies the calculation greatly, that's intentional.
-    const double aim_level = p.get_weapon_dispersion( *weapon ) + predicted_recoil + p.recoil_vehicle();
+    const double dispersion = p.get_weapon_dispersion( *weapon ) + predicted_recoil + p.recoil_vehicle();
     const double range = rl_dist( p.pos(), target->pos() );
-    const double missed_by = aim_level * 0.00021666666666666666 * range;
-    const double hit_rating = missed_by;
-    const double confidence = 1 / hit_rating;
+
+    const double p_headshot = p.projectile_attack_chance( dispersion, range, accuracy_headshot );
```
Comment (mugling) [inline src/ranged.cpp]: Again vertical paging and line changes - this would be much more concise if these variables were moved inside the below calls to `std::make_pair`

File: src/ranged.cpp (hunk 10 above)
Comment (mutability) [inline src/ranged.cpp]: Yes.

File: src/ranged.cpp (hunk 5 above)
Comment (mutability) [inline src/ranged.cpp]: They are not used outside ranged.cpp; do you still want them moved?

File: src/ranged.cpp (hunk 6 above)
Comment (mutability) [inline src/ranged.cpp]: Style comment.

File: src/ranged.cpp (hunk 10 above)
Comment (mugling) [inline src/ranged.cpp]: Is this the change that leads to the smoother graph in 18943? Why does it need to be iterative?

File: src/ranged.cpp (hunk 7 above)
Comment (mutability) [inline src/ranged.cpp]: Style comment.

File: src/creature.h (hunk 1 above)
Comment (mugling) [inline src/creature.h]: `[0..1]` or `(0..1)`. Can you document this here as it could be interpreted as reutrning a percentage chance?

File: src/ranged.cpp (hunk 9 above)
Comment (mutability) [inline src/ranged.cpp]: It's kinda unit tested by definition (a unit test would test that std:erf(z) - x ~ 0 for various values, which is exactly what the core of the function is doing), but fair enough.

File: src/ranged.cpp (hunk 4 above)
Comment (mugling) [inline src/ranged.cpp]: You need to reference magic numbers in code as close to their point of use as possible. In this case a doxygen comment would be ideal

File: src/ranged.cpp (hunk 12 above)
Comment (mutability) [inline src/ranged.cpp]: max_range is used twice.

File: src/ranged.cpp (hunk 13 above)
Comment (mutability) [inline src/ranged.cpp]: Style comment

File: src/ranged.cpp (hunk 5 above)
Comment (mugling) [inline src/ranged.cpp]: Can some of them be moved inside functions? If they can't my gut feeling is that dispersion_sigma belongs in `game_constants.h` as it is after all a game constant.

File: src/ranged.cpp (hunk 11 above)
Comment (mutability) [inline src/ranged.cpp]: max_range is used twice, I'm factoring out the call to gun_range.

File: src/ranged.cpp (hunk 6 above)
Comment (mugling) [inline src/ranged.cpp]: It's inconsistent with the existing codebase

File: src/ranged.cpp (hunk 7 above)
Comment (mugling) [inline src/ranged.cpp]: No it's definitely a spurious comment (pun intended). It's of no value so should be removed

File: src/ranged.cpp (hunk 10 above)
Comment (mutability) [inline src/ranged.cpp]: Yeah, this produces the smoother graph.

There are various ways to compute erfinv but it is a transcendental function so there is no analytic exact way to do it. This is a Newton-Raphson approximation that relies on already having erf. You can also do it with a big higher-order polynomial with lots of magic constants.

For our purposes, if iteration is a problem then the simplest thing here would be to short-circuit the x = 0.5 case and return a constant since that's almost all of the calls.

File: src/ranged.cpp (hunk 8 above)
Comment (mugling) [inline src/ranged.cpp]: That sounds good. It's not essential here (hence tangent) but would be a welcome addition in a separate PR if someone was to chef up a C++17 compliant implementation.

**Comment by mutability** [comment]
For the "style comment" cases I'll do a pass over those once any core issues are sorted out.

File: src/ranged.cpp (hunk 9 above)
Comment (mugling) [inline src/ranged.cpp]: Once your unit test is passing you can then write a non-iterative version? Disclaimer being that I haven't investigated the maths for that yet

File: src/ranged.cpp (hunk 5 above)
Comment (mutability) [inline src/ranged.cpp]: No, they're used by both projectile_attack and projectile_attack_chance.

File: src/ranged.cpp (hunk 12 above)
Comment (mugling) [inline src/ranged.cpp]: But it's only ever calculated once and you haven't added anything here other than line count.

File: src/ranged.cpp (hunk 11 above)
Comment (mugling) [inline src/ranged.cpp]: As above you're factoring out nothing but just increasing line count

**Comment by mugling** [comment]
> For the "style comment" cases I'll do a pass over those once any core issues are sorted out.

Most of those are aimed at reducing line count which helps review so they should be done prior to any further review effort

File: src/ranged.cpp (hunk 13 above)
Comment (mugling) [inline src/ranged.cpp]: As above

**Comment by mutability** [comment]
Well, I'd like to get the core issues sorted out first. Aside from the code review itself, are you happy with the output of the stats dumper etc?

**Comment by mugling** [comment]
> Well, I'd like to get the core issues sorted out first. Aside from the code review itself, are you happy with the output of the stats dumper etc?

No, there's a lot more to do here. Cutting down line count helps so you should do this. Your next steps would also include a unit test for `erfinv`

File: src/ranged.cpp (hunk 9 above)
Comment (mutability) [inline src/ranged.cpp]: You want the giant polynomial? The iterative version is much clearer.

**Comment by mutability** [comment]
Ok, what are the issues with the stats dumper output?

File: src/ranged.cpp (hunk 9 above)
Comment (mugling) [inline src/ranged.cpp]: A black-box unit test would be useful. Not sure about the latter as-yet. I don't want to ask you to write something unless its clear what the benefits are. The graphs are already much smoother so this does suggest your revised version resolves much of the original concern.

File: src/ranged.cpp (hunk 10 above)
Comment (mugling) [inline src/ranged.cpp]: It wouldn't hurt to comment `Newton-Raphson` within the function body

**Comment by mugling** [comment]
Try to close some more of the above review points

File: src/monattack.cpp (hunk 2 above)
Comment (mutability) [inline src/monattack.cpp]: I don't have any particular opinion on what this value should be, what do you suggest it is changed to? Probably something to do in a separate PR though.

File: src/ranged.cpp (hunk 9 above)
Comment (mutability) [inline src/ranged.cpp]: FWIW here is an example of another approach that isn't iterative, I think it's much less clear and for our purposes the current code is fine. Note that with such a large epsilon the loop usually only runs 2-3 times.

```
__inline__ __device__ float MBG_erfinv(float x)
{
float w, p;
w = - __logf((1.0f-x)*(1.0f+x));
if ( w < 5.000000f ) {
w = w - 2.500000f;
p = 2.81022636e-08f;
p = 3.43273939e-07f + p*w;
p = -3.5233877e-06f + p*w;
p = -4.39150654e-06f + p*w;
p = 0.00021858087f + p*w;
p = -0.00125372503f + p*w;
p = -0.00417768164f + p*w;
p = 0.246640727f + p*w;
p = 1.50140941f + p*w;
}
else {
w = sqrtf(w) - 3.000000f;
p = -0.000200214257f;
p = 0.000100950558f + p*w;
p = 0.00134934322f + p*w;
p = -0.00367342844f + p*w;
p = 0.00573950773f + p*w;
p = -0.0076224613f + p*w;
p = 0.00943887047f + p*w;
p = 1.00167406f + p*w;
p = 2.83297682f + p*w;
}
return p*x;
}
```

File: src/ranged.cpp (hunk 11 above)
Comment (mutability) [inline src/ranged.cpp]: OK, fair enough, the paths are independent.

File: src/ranged.cpp (hunk 9 above)
Comment (mugling) [inline src/ranged.cpp]: Is there any difference in accuracy?

File: src/monattack.cpp (hunk 2 above)
Comment (mugling) [inline src/monattack.cpp]: I'm not particularly certain either. Try 250 and see if it's playable?

**Comment by mutability** [comment]
If you could focus on your original objections which were around the stats dumper output not having being looked at, while I look at your review comments in parallel, that'd be great. At the moment none of the review produces functional changes, it's just code cleanup.

**Comment by mugling** [comment]
Reducing line count makes it easier to see what you've changed and therefore makes it easier to review

File: src/ranged.cpp (hunk 9 above)
Comment (mutability) [inline src/ranged.cpp]: The paper I lifted it from suggests it's good for single-precision accuracy. The iterative approach has epsilon=1e-6. Accuracy really isn't a big deal for our purposes here anyway, I doubt you'd see even something as huge as a 1% error.

**Comment by mutability** [comment]
Sure, I can work through that so future code review is easier, but I'm not asking about code review here.

File: src/ranged.cpp (hunk 9 above)
Comment (mutability) [inline src/ranged.cpp]: I've added a unit test that tests within the region we care about.

File: src/creature.h (hunk 1 above)
Comment (mutability) [inline src/creature.h]: This is already a doxygen comment AFAIK and it appears in the generated documentation, what do you mean by "Function comments should be doxygen style".

File: src/monattack.cpp (hunk 2 above)
Comment (mutability) [inline src/monattack.cpp]: Will leave this for a separate PR.

File: src/ranged.cpp (hunk 8 above)
Comment (mutability) [inline src/ranged.cpp]: Will leave this for a separate PR.

File: src/ranged.cpp (hunk 5 above)
Comment (Coolthulhu) [inline src/ranged.cpp]: `game_constants.h` could be a good idea if we don't ever want to change those, but on the other hand those can be seen as implementation details.

I'd prefer them here for as long as the new implementation is not considered fully mature. I'd move them the moment someone needs them outside ranged.cpp (which shouldn't ever be necessary, but may become).

But I wouldn't stress that choice too much - it's mostly a nitpick.

File: src/ranged.cpp (hunk 6 above)
Comment (Coolthulhu) [inline src/ranged.cpp]: You could also use `std::max`. Or even better: clamp `dispersion` value above, at the top of the function. Having negative dispersion could stealthily wreck some things below.

Comment (mutability) [inline src/ranged.cpp]: `std::max` is not appropriate here because `normal_roll` may return negative values. It's expected that `shot_dispersion` might be negative. This conditional is about giving a different meaning to `dispersion <= 0`, normally (heh) it would be meaningless because the normal distribution is undefined for sigma <= 0

File: src/ranged.cpp [hunk 14]
```diff
@@ -929,33 +1035,50 @@ static int do_aim( player &p, const std::vector<Creature *> &t, int cur_target,
...
-        std::make_pair( accuracy_headshot, '*' ),
-        std::make_pair( accuracy_goodhit, '+' ),
-        std::make_pair( accuracy_grazing, '|' ) }};
+        std::make_pair( p_headshot, '*' ),
+        std::make_pair( p_goodhit, '+' ),
+        std::make_pair( p_grazing, '|' ) }};
 
     const int window_width = getmaxx( w ) - 2; // Window width minus borders.
-    const std::string &confidence_bar = get_labeled_bar( confidence, window_width, _( "Confidence" ),
+    const std::string &confidence_bar = get_labeled_bar( 1.0, window_width, _( "Confidence" ),
```
Comment (Coolthulhu) [inline src/ranged.cpp]: 1.0?

Comment (mutability) [inline src/ranged.cpp]: get_labeled_bar puts the transition points at (val * bar_width * value from the iterator).
But the values in the iterators are already the fraction of the bar width I want to use, so there's no use to that extra scale factor.

**Comment by Coolthulhu** [comment]
> This also fixes #19013, #19014, #19015, #19016, #19017, #19018

Pretty sure it has to be in the OP for it to work automatically.

**Comment by Coolthulhu** [comment]
As for the dumper: it produces a large output of weapon stats.
Call it as one of
```
./cataclysm --dump-stats "GUN" > filename.tsv
./cataclysm --dump-stats "GUN" "HTML" > filename.html
```

and then upload it somewhere.

I assume we're trying to match effective ranges here. As long as they're within ~20% of the ones in master, it should be well tolerable, anything more than 30% will need addressing or at least explanation. Outliers can happen, especially among non-firearms.
That is assuming that effective range keeps its original meaning: 50% chance to get a good hit or better at full aim.

File: src/ranged.cpp (hunk 8 above)
Comment (mugling) [inline src/ranged.cpp]: Sure

File: src/monattack.cpp (hunk 2 above)
Comment (mugling) [inline src/monattack.cpp]: Sure

File: src/creature.h (hunk 1 above)
Comment (mugling) [inline src/creature.h]: For example here you should write:
```
/**
 *  Calculate projectile attack to hit with at least the given accuracy
 *  @return chance (0..1)
 */
```

If you use `@return`, `@param`, `@warning` doxyen extracts those separately

**Comment by mugling** [comment]
> and then upload it somewhere.

Use a GH gist

> I assume we're trying to match effective ranges here.

Yes, but I'd prefer either <10% along with a terse explanation of why those values have changed as it was an expected change.

It would be good to look through the code further so please can you complete the outstanding review points

File: src/creature.h (hunk 1 above)
Comment (mugling) [inline src/creature.h]: A lot of the project is poorly documented but this is rapidly improving. Its encouraging that you write documentation for your work so if you could write this in the current style that would be great. There are obviously exceptions where a single line is preferable but ambiguous returns are a great example for the use of `@return`

**Comment by mutability** [comment]
Gist doesn't handle this data well, I'll put it on dropbox or something for your convenience.

Effective and snapshot ranges are 88-89% of the old values. I have verified this in the dump data. This is a direct, expected, consequence of the changes and I reproduce my comment from #18943 below.

tl;dr:

```
>>> import scipy.special
>>> 0.5 / (scipy.special.erfinv(0.50) * (2 ** 0.5) / (2.4 * 0.5))
0.88956133110336111
```

See also the range-for-expected-damage curves at the gist linked below, which is more interesting than comparing "effective range".

---

Effective range is `gun_current_range(gun, dispersion, 50, accuracy_goodhit)`. That is, the range at which you have a 50% chance of getting a hit with an accuracy_goodhit (0.5) or better. (correct me if I'm wrong)

The range is effectively just a scale factor on the accuracy so you can equivalently think of it as "what is the required accuracy at range 1 such that 50% of rolls are closer than that" and then work backward to find the range that gives the requested accuracy. So let's talk about that range-independent bit.

In the old code, it is a little tricky to reason about because of the off-zero distribution of actual dispersions. For the 50% case, it's easy: 50% the distribution is to the left of the mean, so the mean is what you want, and with the old distribution the mean is at exactly half of the total dispersion. This is what effective range ends up using. However, I think other uses of `gun_current_range` (notably `gun_engagement_range(engagement::maximum)` which asks for a 10% chance) get it wrong: the 10% percentile value is not at 10% of the total dispersion, which I think the old `gun_current_range` assumes.

In the new code, the dispersions are zero-centered and you want to find the range (-x, +x) that occurs 50% (or 10%, or whatever) of the time. This can be expressed in terms of `erf` or its inverse. For a 50% chance x is about 0.67 standard deviations: 

```
>>> import math, scipy.special
>>> scipy.special.erfinv(0.50) * (2 ** 0.5)
0.67448975019608171
>>> math.erf(0.67448975019608171 / (2 ** 0.5))
0.49999999999999994
```

and 0.67 standard deviations is about 0.558*dispersion with the current constants. (`0.67 / (dispersion_sigmas * occupied_tile_fraction)`).

i.e. at 50% chance, the spread of shots is about 12% (0.558 / 0.5) wider than the old code. That translates into about a 12% reduction in effective range.

If you wanted to maintain the "effective range" at about the same you could increase dispersion_sigmas from 2.4 to around 2.7, which narrows the distribution, but be aware that this is a direct buff to overall damage. The current value of 2.4 I picked by eye to produce an expected damage curve close to the old one for low-dispersion weapons (high-dispersion weapons get a bit of a buff at longer ranges, they are slightly less useless..). see the first graph at https://gist.github.com/mutability/625b8edb72aceb6abfa161e0acc1c681, the 150 arcmin dispersion one.

This is what I mean about maybe the effective range not being a useful metric here, it only reflects the shape of the center of the distribution and ignores the tail.

**Comment by mutability** [comment]
https://www.dropbox.com/s/v8djnhihy76r21v/gun.tsv?dl=0
https://www.dropbox.com/s/p2sg90oz227j7l2/gun.html?dl=0

File: src/creature.h (hunk 1 above)
Comment (mutability) [inline src/creature.h]: Right, my point was that the current style doesn't match what you're requesting. I'll add the javadoc-style tags, they are good to have, but I'm not psychic and elsewhere you have been unhappy when I have not followed the style of the surrounding code.

**Comment by mutability** [comment]
Only thing left for me to do here is work out what astyle is complaining about in rng.cpp (which will require me building a new astyle, so that might be a while). But that shouldn't halt code review and especially not balance review.

**Comment by mutability** [comment]
Actually the build with the astyle failure includes a failure to fetch the updated source so it may be entirely spurious; can someone with the right astyle version check again?

**Comment by mutability** [comment]
No more changes from me pending, let me know if I missed anything, otherwise I'm waiting on balance feedback or a merge.

File: tests/gun_aiming.cpp [hunk 15]
```diff
@@ -8,19 +8,16 @@ static void test_internal( const npc& who, const item &gun )
...
+            CAPTURE( range );
+            CAPTURE( dispersion );
 
-        INFO( "Recoil: " << recoil );
-        INFO( "Range: " << range );
-        INFO( "Dispersion: " << dispersion );
-
-        // require inverse calculation to agree with tolerance of 0.1%
-        REQUIRE( std::abs( missed_by - accuracy_critical ) < accuracy_critical / 1000 );
+            REQUIRE( who.projectile_attack_chance( dispersion, range, accuracy_critical ) == Approx( 0.5 ).epsilon( 0.0005 ) );
```
Comment (mugling) [inline tests/gun_aiming.cpp]: Dont forget to update `accuracy_critical` here or even better define both this and the chance as separate variables and then iterate through different values for chance in an outer loop.

I suspect your implementation would pass such a test whereas your comment with the math reasoning in it suggests the current implementation would not (would fail for `chance != 50`)

**Comment by mugling** [comment]
Excluding those changes to the unit tests the only other thing to do here is to check for value differences.

We could either do acceptance tests (preferable but laborious) or compare the output of dump-stats for effective and maximum range. You could script calls to `--dump-stats` and then rank items by percentage change and post either your script or the top/bottom 5% (preferably both)?

File: tests/gun_aiming.cpp [hunk 16]
```diff
@@ -8,19 +8,16 @@ static void test_internal( const npc& who, const item &gun )
 {
     THEN( "the effective range is correctly calcuated" ) {
         // calculate range for 50% chance of critical hit at arbitrary recoil
-        double recoil = rng_float( 0, 1000 );
-        double range = who.gun_current_range( gun, recoil, 50, accuracy_critical );
+        for ( double recoil = 0; recoil < 1000; recoil += 100 ) {
+            double range = who.gun_current_range( gun, recoil, 50, accuracy_critical );
```
Comment (mutability) [inline tests/gun_aiming.cpp]: I'll change it to use gun_engagement_range, then it'll catch any changes in there too.

Comment (mutability) [inline tests/gun_aiming.cpp]: Oh, except that uses recoil, never mind.

Comment (mugling) [inline tests/gun_aiming.cpp]: Yes I looked at that solution first but you can only pass base recoil. A solution with nested loops checking various values of `recoil` and `chance` would be best

File: src/ranged.cpp [hunk 17]
```diff
@@ -929,33 +991,46 @@ static int do_aim( player &p, const std::vector<Creature *> &t, int cur_target,
     return cur_target;
 }
 
+double Creature::projectile_attack_chance( double dispersion, double range, double accuracy ) const {
+    // This is essentially the inverse of what Creature::projectile_attack() does.
```
Comment (mugling) [inline src/ranged.cpp]: sounds good. I think if you do this here and extend the other tests as discussed we should have fairly comprehensive coverage

File: src/creature.h [hunk 18]
```diff
@@ -39,6 +39,25 @@ enum m_size : int {
     MS_HUGE     // TAAAANK
 };
 
+/**
+ * Aim result for a single projectile attack
+ */
+struct projectile_attack_aim {
```
Comment (mugling) [inline src/creature.h]: Not sure about the name but I don't currently have a better suggestion

File: src/creature.h [hunk 19]
```diff
@@ -186,6 +205,15 @@ class Creature
                                                    double total_dispersion );
 
         /**
+         * Makes an aiming/attack roll for a single projectile attack shot.
+         *
```
Comment (mugling) [inline src/creature.h]: Style: no line break between preamble and following lines. Also no stop at end of preamble sentence

File: src/creature.h [hunk 20]
```diff
@@ -186,6 +205,15 @@ class Creature
...
+         * @param dispersion the base dispersion of the projectile attack
+         * @param range the range of the projectile attack
+         * @return the results of the aiming roll
```
Comment (mugling) [inline src/creature.h]: The extra definitions here don't add much. Doxygen comments aren't obligatory when they don't add any relevant information

File: src/creature.h (hunk 19 above)
Comment (mutability) [inline src/creature.h]: Can you point me at the doxygen style guidelines?

**Comment by mutability** [comment]
A note on the gun_aiming test (this is a pre-existing problem): it only tests the first gun it finds, despite what you'd expect from a cursory look at the code.

File: src/creature.h (hunk 19 above)
Comment (mugling) [inline src/creature.h]: Sorry about that. I guess it's just a case of getting used to it really. Consistency is important but the fact that you write documentation at all is welcome (compared to a lot of the legacy codebase which is entirely undocumented)

Comment (mutability) [inline src/creature.h]: doxygen_conf.txt has JAVADOC_AUTOBRIEF = YES. If you want "no stop at the end of preamble sentence" then you should set that to NO; the current setting relies on the stop to find the end of the brief description.

Comment (mugling) [inline src/creature.h]: Except where `@param`, `@return` etc follow, which they usually do

**Comment by mutability** [comment]
I pulled the old vs new stats dump into a spreadsheet and did the comparison by hand.

No changes to calculated maximum ranges at all (it's a bit of a useless statistic, all guns cap out at the gun's maximum range or MAX_RANGE before getting anywhere near the underlying calculated value)

Largest change to effective range: 87.5% of original (spearguns 2.4->2.1, nail rifle 4.8->4.2)

At the other end of the scale are a few guns where the effective range was previously larger than the gun max range, and now isn't:
 * Boeing XM-P plasma rifle 12 (capped) -> 11.8 (98.3%)
 * pistol crossbow 6 (capped) -> 5.8 (96.7%)
 * manual autoshotgun 6 (capped) -> 5.5 (91.7%)

Everything else is either no change (was capped at max range before, still is now) or between 88%-91% of the old effective range. The variation is mostly just rounding errors since the effective range is only printed with 1dp.

Snapshot ranges follow the same pattern with somewhat more variation as they are shorter ranges and so  more affected by the rounding.

 * pneumatic speargun 1.9 -> 1.6 (84.2%)
 * others 86 - 91%
 * combination gun shotgun 4.9 -> 4.5 (91.8%)
 * spraycan flamethrower is the only one with an entirely unchanged snapshot range (3, capped at max range)

Largest absolute change in effective range was RM11B scout rifle (20.8 -> 18.5, -2.3 tiles, 88.9%) followed unsurprisingly by a bunch of other long range rifles and energy weapons.

Overall seems consistent with the 89% value predicted above.

File: src/creature.h (hunk 19 above)
Comment (mutability) [inline src/creature.h]: You should still change it, if only as the start of some documentation of what you want.

**Comment by mutability** [comment]
Oh and the H&K G80 Railgun produce zeros for all ranges in both versions for reasons I haven't looked into.

**Comment by mugling** [comment]
Looks good

> A note on the gun_aiming test (this is a pre-existing problem): it only tests the first gun it finds, despite what you'd expect from a cursory look at the code.

How so? I remember having a lot of problems getting that unit test to work with the loop

**Comment by mutability** [comment]
The basic problem is that Catch will only execute a leaf test (WHEN, THEN, GIVEN) exactly once, regardless of how it's nested within loops and functions calls.

The simplest way to see it is to drop a debug printf in one of the THENs in test_internal - it'll only get run once, with the 2-shot special.

Fixing it is a bit horrible because you have to put the looping over gun types within each WHEN/THEN/etc or forgo the WHEN/THEN/etc entirely and lose some isolation between the leaf tests.

**Comment by mugling** [comment]
Needs trivial rebase but I the `[aim]` unit tests are all passing

**Comment by mugling** [comment]
You are correct though that it only tests for the first item. Can you pick a few representative items (say `glock19`, `mp5`, `ar15` and `shotgun`)?

**Comment by mutability** [comment]
I moved the loops inside the leaf tests, if you prepare a vector of test guns ahead of time it's not so bad.

**Comment by mugling** [comment]
This is a high quality implementation complete with inline documentation and backed by unit tests. I'd prefer it didn't reduce the ranges but overall the effect on gameplay is positive due to the more sane handling of misses. In particular the idea about `occupied_tile` proportion is interesting.

Minor: `Creature::projectile_attack_roll` doesn't need to be a member function (we are trying to reduce size of many core classes).
